### PR TITLE
#1494 Correct switchOff handle of connect back an equipment. Better h…

### PR DESCRIPTION
**Pull request by gautierbureau**
…andle of locked/running in TapChanger.

### CODE REVIEW

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Basics/SwitchOff.mo [hunk 1]
```diff
@@ -35,14 +35,20 @@ partial model SwitchOffLogic "Manage switch-off logic"
     if (NbSwitchOffSignals >= 3) then
       when switchOffSignal1.value or switchOffSignal2.value or switchOffSignal3.value and pre(running.value) then
         running.value = false;
+      elsewhen not switchOffSignal1.value and not switchOffSignal2.value and not switchOffSignal3.value and not pre(running.value) then
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Basics/SwitchOff.mo:38]: done

File: nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.dyd [hunk 2]
```diff
@@ -0,0 +1,32 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<!--
+    Copyright (c) 2015-2019, RTE (http://www.rte-france.com)
```
Comment (rosiereflo) [inline nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.dyd:3]: 2021 (same for all files)

File: nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.jobs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, you can obtain one at http://mozilla.org/MPL/2.0/.
+    SPDX-License-Identifier: MPL-2.0
+
+    This file is part of Dynawo, an hybrid C++/Modelica open source time domain
+    simulation tool for power systems.
+-->
+<dyn:jobs xmlns:dyn="http://www.rte-france.com/dynawo">
+  <dyn:job name="Disconnect load test">
```
Comment (rosiereflo) [inline nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.jobs:15]: Let's add the fact that we reconnect in the jobs description

File: nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.dyd (hunk 2 above)
Comment (rosiereflo) [reply nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.dyd:3]: Maybe it should go into example?

File: nrt/data/DisconnectConnectLoad/curves.crv [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, you can obtain one at http://mozilla.org/MPL/2.0/.
+    SPDX-License-Identifier: MPL-2.0
+
+    This file is part of Dynawo, an hybrid C++/Modelica open source time domain
+    simulation tool for power systems.
+-->
+<curvesInput xmlns="http://www.rte-france.com/dynawo">
+  <curve model="Generator" variable="generator_UPu"/>
+  <curve model="Generator" variable="generator_running"/>
```
Comment (rosiereflo) [inline nrt/data/DisconnectConnectLoad/curves.crv:16]: reference are missing

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.dyd (hunk 2 above)
Comment (gautierbureau) [reply nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.dyd]: 2021 done

Comment (gautierbureau) [reply nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.dyd]: for example I don't have an opinion

File: nrt/data/DisconnectConnectLoad/curves.crv (hunk 4 above)
Comment (gautierbureau) [reply nrt/data/DisconnectConnectLoad/curves.crv:16]: done

File: nrt/data/DisconnectConnectLoad/reference/outputs/timeLine/timeline.log [hunk 5]
```diff
@@ -0,0 +1 @@
+30 | Load | LOAD : disconnecting
```
Comment (rosiereflo) [inline nrt/data/DisconnectConnectLoad/reference/outputs/timeLine/timeline.log:1]: Maybe we should add an event for the reconnection?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.jobs (hunk 3 above)
Comment (gautierbureau) [reply nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.jobs]: done

File: nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.dyd (hunk 2 above)
Comment (MarianneSaugierRTE) [reply nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.dyd]: Usually this kind of test goes in dynawo-rte/nrt/data/standaloneCases

Comment (rosiereflo) [reply nrt/data/DisconnectConnectLoad/DisconnectConnectLoad.dyd]: sounds good to me

**Review by MarianneSaugierRTE — APPROVED** [review]

**Comment by gautierbureau** [comment]
I added a timeline when reconnecting an equipment.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Transformers/BaseClasses.mo [hunk 6]
```diff
@@ -201,26 +201,26 @@ partial model BaseTapChangerPhaseShifter_INTERVAL "Base model for tap-changers a
...
     tValueUnderMinWhileRunning = pre(tValueUnderMinWhileRunning);
     valueAboveMax = false;
     tValueAboveMaxWhileRunning = pre(tValueAboveMaxWhileRunning);
-  elsewhen running.value and locked then
+  elsewhen (running.value and locked) or not(running.value) then
     valueUnderMin = pre(valueUnderMin);
     tValueUnderMinWhileRunning = Constants.inf;
     valueAboveMax = pre(valueAboveMax);
-    tValueAboveMaxWhileRunning = pre(tValueAboveMaxWhileRunning);
+    tValueAboveMaxWhileRunning = Constants.inf;
```
[the remark this answers is not in the document]
Comment (MarianneSaugierRTE) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Transformers/BaseClasses.mo:223]: It was a mistake ;)

[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Transformers/BaseClasses.mo:223]: So if I get you well this new behavior is the correct one?

[the remark this answers is not in the document]
Comment (MarianneSaugierRTE) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Transformers/BaseClasses.mo:223]: Yes, sorry for the confusing sentence

**Review by rosiereflo — APPROVED** [review]

**Comment by rosiereflo** [comment]
@adriguir @MarianneSaugierRTE is there still some open points on this PR?

**Review by MarianneSaugierRTE — APPROVED** [review]

**Comment by adriguir** [comment]
> @adriguir @MarianneSaugierRTE is there still some open points on this PR?

I haven't gone in details through it but it seems fine for me. I will launch the update. Out of curiosity, any impact on the performances? (as the switch off condition is part of each Modelica model by default, it should slow down a bit the g and z evaluations I suppose). It won't be a criterion not to integrate but could be good to know and to have in mind how much this kind of things impact the performances.

**Review by adriguir — APPROVED** [review]

**Pull request merged**